diff -up chrony-3.1/examples/chronyd.service.service-helper chrony-3.1/examples/chronyd.service
--- chrony-3.1/examples/chronyd.service.service-helper	2017-01-31 12:12:01.863772826 +0100
+++ chrony-3.1/examples/chronyd.service	2017-01-31 12:12:30.371860064 +0100
@@ -10,6 +10,7 @@ Type=forking
 PIDFile=/var/run/chronyd.pid
 EnvironmentFile=-/etc/sysconfig/chronyd
 ExecStart=/usr/sbin/chronyd $OPTIONS
+ExecStartPost=/usr/libexec/chrony-helper update-daemon
 PrivateTmp=yes
 ProtectHome=yes
 ProtectSystem=full
