diff -up chrony-4.1/examples/chronyd.service.service-helper chrony-4.1/examples/chronyd.service
--- chrony-4.1/examples/chronyd.service.service-helper	2021-05-12 13:06:15.000000000 +0200
+++ chrony-4.1/examples/chronyd.service	2021-06-15 09:01:56.948968576 +0200
@@ -10,6 +10,8 @@ Type=forking
 PIDFile=/run/chrony/chronyd.pid
 EnvironmentFile=-/etc/sysconfig/chronyd
 ExecStart=/usr/sbin/chronyd $OPTIONS
+ExecStartPost=/usr/libexec/chrony-helper update-daemon
+ExecStopPost=/usr/libexec/chrony-helper remove-daemon-state
 PrivateTmp=yes
 ProtectHome=yes
 ProtectSystem=full
